Logic: update `planner_move` to handle co-requisites

Letm3through opened the pull request:
Resolves #164. 

The application allows the user to set co-requisites to modules. However, `planner_move` does not handle co-requisites. Hence, the user has to move the codes that shares co-requisites manually.

Let's enhance `planner_move` command by allowing it to handle co-requisites so that when a module is moved, its co-requisites are moved along if they exist in the degree plan.

In addition, since there has been update on `ModuleList` to store the semesters the module is offered in, let's also enhance `planner_move` command to check if the module is offered in the semester the user wants to move to.

The followings are the changes in the process of the enhancement:
* [1/3] [Logic: update 'planner_move' to handle co-requisites](https://github.com/CS2113-AY1819S2-T09-1/main/pull/197/commits/1bc4468057d89578f5e5b58a207ad25b40a0b020)
* [2/3] [Model: add method to help in carrying out handling of co-requisites](https://github.com/CS2113-AY1819S2-T09-1/main/pull/197/commits/181b24e3fcb847d3c7485740bb4ecd47652060e0)
* [3/3] [Test: update test cases for 'planner_move' command](https://github.com/CS2113-AY1819S2-T09-1/main/pull/197/commits/ed22a52fa272b0c623ca762922e9932dbc4370e7)

Letm3through committed:
Logic: update 'planner_move' to handle co-requisites

The application allows the user to set co-requisites to modules.
However, 'planner_move' does not handle co-requisites.
Hence, the user has to move the codes that shares co-requisites
manually.

Let's enhance 'planner_move' command by allowing it to handle
co-requisites so that when a module is moved, its co-requisites are
moved along if they exist in the degree plan.

lycjackie commented on src/main/java/pwe/planner/logic/commands/PlannerMoveCommand.java:
```diff
@@ -93,8 +103,14 @@ public CommandResult execute(Model model, CommandHistory history) throws Command
+        StringBuilder codesMoved = new StringBuilder();
+        codesMoved.append(toMove).append(" to Year ").append(destinationYear).append(" Semester ")
+                .append(destinationSemester).append(" of the degree plan!\nCo-Requisite(s) moved along: ")
+                .append(corequisitesToMove.stream().map(Code::toString).collect(Collectors.joining(", ")));
```
Perhaps, instead of using a `StringBuilder` here, you might want to consider the following?
```
MESSAGE_SUCCESS = "Successfully moved %1$s to Year %2$s, Semester %3$s of degree planner\n"
                        + "Co-Requisite(s) moved along: %4$s"
coReqMoved = corequisitesToMove.stream().map(Code::toString).collect(Collectors.joining(", "));

String.format(MESSAGE_SUCCESS, toMove, destinationYear, destinationSemester, coRedMoved);
```

lycjackie commented on src/main/java/pwe/planner/logic/commands/PlannerMoveCommand.java:
```diff
@@ -93,8 +103,14 @@ public CommandResult execute(Model model, CommandHistory history) throws Command
         model.commitApplication();
-        return new CommandResult(String.format(MESSAGE_SUCCESS, toMove, destinationYear, destinationSemester));
+        return new CommandResult(
+                String.format(MESSAGE_SUCCESS, codesMoved.toString()));
```
Do you think you can change this to 1 line? This line looks shorter than the previous thou.

No reason to split it

lycjackie commented on src/test/java/pwe/planner/logic/commands/PlannerMoveCommandTest.java:
```diff
@@ -84,18 +86,24 @@ public void execute_validPlannerUnfilteredList_success() throws Exception {
                 .orElse(null);
 
+        Set<Code> allDegreePlannerCodes =
+                model.getApplication().getDegreePlannerList().stream()
```
Can you join this line with L89? It does not look like it will reach the line limit.

lycjackie commented on src/test/java/pwe/planner/logic/commands/PlannerMoveCommandTest.java:
```diff
@@ -104,6 +112,11 @@ public void execute_validPlannerUnfilteredList_success() throws Exception {
         expectedModel.setDegreePlanner(destinationPlannerToEdit, editedDestinationPlanner);
+
+        String expectedMessage =
+                String.format(PlannerMoveCommand.MESSAGE_SUCCESS,
```
Same as above comment

lycjackie commented on src/test/java/pwe/planner/logic/commands/PlannerMoveCommandTest.java:
```diff
@@ -123,14 +136,21 @@ public void execute_moveToSamePlannerUnfilteredList_success() {
         Year validYear = new Year("1");
         Semester validSemester = new Semester("1");
+        Set<Code> allDegreePlannerCodes =
+                model.getApplication().getDegreePlannerList().stream()
```
Same as above comment

lycjackie commented on src/test/java/pwe/planner/logic/commands/PlannerMoveCommandTest.java:
```diff
@@ -188,4 +208,14 @@ public void equals() {
+    /**
+     * @return String that is to be passed on the success of executing {@code PlannerMoveCommand}
+     */
+    private String codesMovedContent(Code code, Year year, Semester semester, Set<Code> corequisites) {
```
Maybe if you used the format string style, this method could be reduced accordingly?

Letm3through committed:
Model: add method to help in carrying out handling of co-requisites

Let's add a method 'moveModuleBetweenPlanner' that takes in 'code'
to allow 'planner_move' command to handle co-requisites.

Letm3through committed:
Test: update test cases for 'planner_move' command

There has been changes to 'planner_move' command in the process of
enhancement to handle co-requisite.

Let's update test cases for 'planner_move' command to suite to
the changes.

Note: 'TypicalModules' and 'typicalModulesList' are modified to
allow more test for 'planner_move'.

lycjackie reviewed (commented):
LGTM,

I will be tagging @Creastery on the Model part. 

Great work!

Creastery commented on src/main/java/pwe/planner/logic/commands/PlannerMoveCommand.java:
```diff
@@ -25,7 +26,8 @@
 
-    public static final String MESSAGE_USAGE = COMMAND_WORD + ": Moves a module in a degree plan. "
+    public static final String MESSAGE_USAGE = COMMAND_WORD + ": Moves a module along with its co-requisites "
+            + "in a degree plan.\n"
```
Maybe it would be good to mention "move ... to another year/semester of the degree plan".

Creastery commented on src/main/java/pwe/planner/logic/commands/PlannerMoveCommand.java:
```diff
@@ -35,12 +37,12 @@
-    public static final String MESSAGE_SUCCESS =
-            "Successfully moved %1$s to Year %2$s Semester %3$s of the degree plan!";
+    public static final String MESSAGE_SUCCESS = "You have successfully moved %1$s to Year %2$s, Semester %3$s of "
+            + "degree planner!\n" + "Co-Requisite(s) moved along: %4$s";
```
Let's standardise the success message to be:
"Successfully moved ...."

Creastery replied:
If the module has no co-requisites, `%4$s` will be blank as you didn't handle a blank string for `%4$s` below.

Creastery commented on src/main/java/pwe/planner/model/Application.java:
```diff
@@ -328,6 +328,30 @@ public void setDegreePlanner(DegreePlanner target, DegreePlanner editedDegreePla
+
+        Set<Code> corequisitesCode = modules.getModuleByCode(code).getCorequisites();
+        Set<Code> editedCodes = new HashSet<>(editedDegreePlanner.getCodes());
+        if (target.getCodes().size() > editedDegreePlanner.getCodes().size()) {
```
Why not just check if `target.getCodes().equals(editedDegreePlanner.getCodes())`?

Letm3through replied:
Thank you for the suggesting the improvement that can be made. Changes made.

Creastery commented on src/test/java/pwe/planner/logic/commands/PlannerMoveCommandTest.java:
```diff
@@ -102,8 +97,17 @@ public void execute_validPlannerUnfilteredList_success() throws Exception {
+                .collect(Collectors.joining(", "));
+        String expectedMessage =
+                String.format(PlannerMoveCommand.MESSAGE_SUCCESS, validCodeToMove, validYear, validSemester,
+                        corequisitesMoved);
```
Make this statement 2 lines instead.

Creastery commented on src/test/java/pwe/planner/logic/commands/PlannerMoveCommandTest.java:
```diff
@@ -143,13 +148,45 @@ public void execute_moveToSamePlannerUnfilteredList_success() {
+
+        assertCommandFailure(new PlannerMoveCommand(validYear, validSemester, unavailableValidCodeToMove), model,
+                commandHistory,
+                String.format(PlannerMoveCommand.MESSAGE_UNAVAILABLE_SEMESTER, validSemester));
```
Here, you are supplying the expected message, which is an error message.
Yet, Line 175 suggests that this test is supposed to be successful!

Creastery commented on src/test/java/pwe/planner/logic/commands/PlannerMoveCommandTest.java:
```diff
@@ -143,13 +148,45 @@ public void execute_moveToSamePlannerUnfilteredList_success() {
+                String.format(PlannerMoveCommand.MESSAGE_SUCCESS, unavailableValidCodeToMove, validYear,
+                        validSemester, corequisitesMoved);
+
+        assertCommandFailure(new PlannerMoveCommand(validYear, validSemester, unavailableValidCodeToMove), model,
```
You didn't use `plannerMoveCommand` variable as first argument to `assertCommandFailure`!

Creastery commented on src/main/java/pwe/planner/logic/commands/PlannerMoveCommand.java:
```diff
@@ -72,29 +76,29 @@ public CommandResult execute(Model model, CommandHistory history) throws Command
-            newSourceCodes.remove(toMove);
-            Set<Code> newDestinationCodes = new HashSet<>(destinationPlanner.getCodes());
-            newDestinationCodes.add(toMove);
+        model.moveModuleBetweenPlanner(sourcePlanner, destinationPlanner, toMove);
```
You didn't check if all the co-requisites to be moved have the target semester offered before you call this method!

Creastery commented on src/main/java/pwe/planner/model/Application.java:
```diff
@@ -352,6 +353,44 @@ public void setDegreePlanner(DegreePlanner target, DegreePlanner editedDegreePla
+        Set<Code> codesToMove = modules.getModuleByCode(code).getCorequisites().stream()
+                .filter(corequisite -> getModuleByCode(corequisite).getSemesters()
+                        .contains(destinationPlanner.getSemester()))
+                .collect(Collectors.toSet());
```
This means you are ignoring co-requisite modules that are not offered the semester you are trying to move the modules to.

But, if the modules are co-requisites, they all need to be offered in the same semester.
In other words, if a module has a co-requisite that is not offered in the semester to move to, all these modules cannot be moved to the same semester (all are co-requisites of each other).

Creastery commented on src/main/java/pwe/planner/logic/commands/PlannerMoveCommand.java:
```diff
@@ -35,12 +38,27 @@
+
+    public static final String MESSAGE_UNAVAILABLE_SEMESTER =
+            "You cannot move a module that is not offered in Semester %1$s"
+            + "Perhaps you may want to check the module list for the semester(s) the module is offered in.";
```
There should be `.\n` after `%1$s`.

Letm3through replied:
Thanks for pointing!

Creastery commented on src/main/java/pwe/planner/logic/commands/PlannerMoveCommand.java:
```diff
@@ -35,12 +38,27 @@
+            + "Perhaps you may want to check the module list for the semester(s) the module is offered in.";
+
+    public static final String MESSAGE_UNAVAILABLE_COREQUISITES =
+            "You cannot move a module that has co-requisite(s) %1$s not offered in Semester %2$s"
```
There should be `.\n` after `%2$s`.

```
that has co-requisite(s) %1$s not offered
```
can be rephrased to:
```
that has co-requisite(s) %1$s, which isn't not offered
```

Letm3through replied:
> There should be `.\n` after `%2$s`.
> 
> ```
> that has co-requisite(s) %1$s not offered
> ```
> can be rephrased to:
> 
> ```
> that has co-requisite(s) %1$s, which isn't not offered
> ```
Thanks for the suggestion. I have changed as below accordingly since I believe `not` here is added unexpectedly.
```
that has co-requisite(s) %1$s, which isn't offered
```

Creastery commented on src/test/java/pwe/planner/testutil/TypicalModules.java:
```diff
@@ -70,7 +70,7 @@
             .withName("Fiona Kunz")
             .withCredits("5")
-            .withSemesters("1", "2")
+            .withSemesters("1")
```
What's the rationale for this change?

Letm3through replied:
This change is have been made to include the test case where module is available on Semester 2 but co-requisite is not and output message for that case accordingly.

Creastery commented on src/main/java/pwe/planner/logic/commands/PlannerMoveCommand.java:
```diff
@@ -35,12 +38,33 @@
+            "You cannot move a module that is not offered in Semester %1$s\n"
+            + "Perhaps you may want to check the module list for the semester(s) the module is offered in.\n"
+            + "[Tip] You may want to refer to the degree plan to see the academic semesters"
+            + " available in the application!";
```
How does looking at the semesters available in the application help the user to be able to move the module to the Semester?

Creastery commented on src/main/java/pwe/planner/logic/commands/PlannerMoveCommand.java:
```diff
@@ -35,12 +38,33 @@
+            "You cannot move a module as Year %1$s Semester %2$s does not exist in the degree plan!\n"
+            + "Perhaps you mistyped either year or semester?\n"
+            + "[Tip] You may want to refer to the degree plan to see the academic semesters"
+            + " available in the application!";
```
How?

Creastery reviewed (approved):
Please see minor changes that needs to be done below. Thanks!

Creastery commented on src/main/java/pwe/planner/model/Application.java:
```diff
@@ -352,6 +352,41 @@ public void setDegreePlanner(DegreePlanner target, DegreePlanner editedDegreePla
+                    new DegreePlanner(sourceDegreePlanner.getYear(), sourceDegreePlanner.getSemester(),
+                            editedSourceDegreePlanner);
+
+            degreePlanners.setDegreePlanner(sourceDegreePlanner, editedSourcePlanner);
```
You can remove `degreePlanners` here to utilise `Application#setDegreePlanner(...)` instead!

Creastery commented on src/main/java/pwe/planner/model/Application.java:
```diff
@@ -352,6 +352,41 @@ public void setDegreePlanner(DegreePlanner target, DegreePlanner editedDegreePla
+                .filter(destinationPlanner::isSameDegreePlanner)
+                .findFirst()
+                .orElse(null);
+        degreePlanners.setDegreePlanner(targetDestinationPlanner, editedDestinationPlanner);
```
You can remove `degreePlanners` here to utilise `Application#setDegreePlanner(...)` instead!

Creastery commented:
LGTM. Approving and merging.
